Chokes on repo components that don't have any packages in them

Eg, `xenial-updates/main/binary-amd64/Packages.gz` doesn't have any packages in it yet.

The error is:

```
object.Exception@source/archive.d(84): Unable to open compressed file '.../Packages.gz'
```

FYI, I'm not actually running this on the real Ubuntu repo. I just gave that as an example. We have a company-specific repo that I'm trying to generate data for. I've had to remove the empty components from `asgen-config.json` temporarily.

For context, I'm on Debian Stretch, without `libappstream-dev` installed, and a copy of https://github.com/ximion/appstream built from `57917bd` installed to `/opt/appstream`. I then run it with:

```
LD_LIBRARY_PATH=/opt/appstream/lib build/appstream-generator process ...
```

I used `ldd` to check that it's picking up the correct copy of `libappstream.so.3`.

I modified `dub.json` to enable it to pick up the locally-built `libappstream` during the build of `asgen`:

```Diff
diff --git a/dub.json b/dub.json
index 52dac6b..3ca1a39 100644
--- a/dub.json
+++ b/dub.json
@@ -28,7 +28,7 @@
     "x-sourcePaths": ["source", "build/bindings"],
     "x-importPaths": ["build/bindings"],
 
-    "dflags": ["-Wl,--push-state,-no-as-needed", "-lcurl", "-Wl,--pop-state"],
+    "dflags": ["-Wl,--push-state,-no-as-needed", "-lcurl", "-Wl,--pop-state", "-I/opt/appstream/include/", "-Wl,-L/opt/appstream/lib/"],
     "systemDependencies": "LMDB, AppStream, GLib2, GIO, GObject, libarchive, GDLib, LibRSVG2",
     "libs": ["lmdb", "archive", "glib-2.0", "gobject-2.0", "gio-2.0", "appstream", "cairo", "gdk-pixbuf-2.0", "librsvg-2.0"]
 }
```

I tried doing it by setting `PKG_CONFIG_PATH` but for some reason `gdc` doesn't pass the `-L` option to `ld` properly and I get link errors (`undefined reference to 'as_metadata_set_architecture'`).

--- a/source/archive.d
+++ b/source/archive.d
@@ -44,6 +44,10 @@ private string readArchiveData (archive *ar, string name = null)
     char[BUFFER_SIZE] buff;
 
     ret = archive_read_next_header (ar, &ae);
+
+    if (ret == ARCHIVE_EOF)
+        return data;
+
     if (ret != ARCHIVE_OK) {
         if (name is null)
             throw new Exception (format ("Unable to read header of compressed data."));
@@ -77,6 +81,7 @@ string decompressFile (string fname)
     scope(exit) archive_read_free (ar);
 
     archive_read_support_format_raw (ar);
+    archive_read_support_format_empty (ar);
     archive_read_support_filter_all (ar);
 
     ret = archive_read_open_filename (ar, toStringz (fname), 16384);
@@ -94,6 +99,7 @@ string decompressData (ubyte[] data)
     scope(exit) archive_read_free (ar);
 
     archive_read_support_filter_all (ar);
+    archive_read_support_format_empty (ar);
     archive_read_support_format_raw (ar);
 
     auto dSize = ubyte.sizeof * data.length;
@@ -465,4 +471,17 @@ public:
         }
     }
 
+}
+
+unittest
+{
+    writeln ("TEST: ", "Compressed empty file");
+
+    ubyte emptyGz [] = [
+       0x1f, 0x8b, 0x08, 0x08, 0x00, 0x00, 0x00, 0x00,
+       0x00, 0x03, 0x65, 0x6d, 0x70, 0x74, 0x79, 0x00,
+       0x03, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+       0x00, 0x00,
+    ];
+    assert (decompressData (emptyGz) == "");
 }

--- a/source/backends/debian/tagfile.d
+++ b/source/backends/debian/tagfile.d
@@ -46,12 +46,7 @@ public:
         try {
             data = decompressFile (fname);
         } catch (Exception e) {
-            // libarchive fails to detect GZip-compressed zero-byte documents proprly and emits an error.
-            // We don't want this to be a permanent failure, so we ignore errors in that particular case.
-            if (fname.endsWith (".gz"))
-                logWarning (e.msg);
-            else
-                throw e;
+            throw e;
         }
 
         content = splitLines (data);

--- a/source/bindings/libarchive.d
+++ b/source/bindings/libarchive.d
@@ -47,6 +47,7 @@ int archive_read_support_filter_gzip (archive*);
 int archive_read_support_filter_lzma (archive*);
 
 int archive_read_support_format_raw (archive*);
+int archive_read_support_format_empty (archive*);
 int archive_read_support_format_all (archive*);
 int archive_read_support_format_ar (archive*);
 int archive_read_support_format_gnutar (archive*);

